### PR TITLE
ENH: Implement hash, eq, neq

### PR DESCRIPTION
**Pull request by caspervdw**
This implement hashing for geometry objects.

### CODE REVIEW

File: src/pygeom.c [hunk 1]
```diff
@@ -76,6 +76,69 @@ static PyObject *GeometryObject_str(GeometryObject *self)
...
+{
+    void *context = geos_context[0];
+    unsigned char *wkb;
+    size_t size;
+    Py_hash_t x;
+
+    if (self->ptr == NULL) {
+        return -1;
+    }
+    GEOSWKBWriter *writer = GEOSWKBWriter_create_r(context);
```
Comment (brendan-ward) [inline src/pygeom.c]: @caspervdw what about using [GEOSGeomToWKB_buf_r](https://github.com/libgeos/geos/blob/3.8/capi/geos_ts_c.cpp#L1370)?  It seems like that would avoid needing to manage the writer yourself.

Example [usage](https://github.com/libgeos/geos/blob/3.8/tests/thread/threadtest.c#L222-L223).

Comment (jorisvandenbossche) [reply src/pygeom.c]: According to the C API header that is function is deprecated:

https://github.com/libgeos/geos/blob/c01c8c1c30091d655ba888790c2daeec1f33bae3/capi/geos_c.h.in#L220

(but not sure how/if this is enforced)

File: src/pygeom.c [hunk 2]
```diff
@@ -76,6 +76,69 @@ static PyObject *GeometryObject_str(GeometryObject *self)
...
+    wkb = GEOSWKBWriter_write_r(context, writer, self->ptr, &size);
+    GEOSWKBWriter_destroy_r(context, writer);
+    if (wkb == NULL) {
+        return -1;
+    }
+    x = PyHash_GetFuncDef()->hash(wkb, size);
+    if (x == -1) {
+        x = -2;
+    } else {
+        x ^= 374761393UL;  // to make the result distinct from the actual WKB hash //
```
Comment (jorisvandenbossche) [inline src/pygeom.c]: is this a random number? Or does it come from somewhere?

Comment (caspervdw) [reply src/pygeom.c]: It is a prime that I stole from some other hash function in the Python source. I guess it could be any number if you use it in a XOR like I did.

**Comment by caspervdw** [comment]
Thanks for the feedback. I compared 'my'  equals implementation with GEOS' equals_exact, and of course, the latter is a lot more efficient:

Going through WKB:
```In [1]: import numpy as np
In [2]: import pygeos
In [3]: a = pygeos.linestrings(np.arange(1000), np.arange(0, 2000, 2))
In [4]: b = pygeos.linestrings(np.arange(1000), np.arange(0, 2000, 2))
In [5]: %timeit a == b
82.6 µs ± 136 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
Using EqualsExact:
```
In [8]: %timeit a == b
6.32 µs ± 34.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

File: src/pygeom.c (hunk 1 above)
Comment (caspervdw) [reply src/pygeom.c]: I think it is no good to use deprecated functions.

**Comment by caspervdw** [comment]
The test failure is unrelated (see #100)

**Comment by jorisvandenbossche** [comment]
On the other hand, for scalar equality, we maybe don't care much about performance? 
When working with arrays, `pygeos.equals_exact(a, b)` will always be faster than `a == b` (even if the scalar comparison itself is as fast as possible)

**Comment by jorisvandenbossche** [comment]
> One concern with EqualsExact might be whether we are sure this will always give the same result as hashing of WKB (to ensure equivalence between hash and eq). But I suppose this will be the case with tolerance of 0.

Maybe it would be good to check come corner cases: empty geometries, geometries with NaN coordinates.

**Comment by caspervdw** [comment]
I added some tests and hit upon a peculiar bit of logic:
```
> set([float("nan"), float("nan")])
{nan, nan}
> a = float("nan"); set([a, a])
{nan}
```

Same can be done for `np.nan`. It has to do with the fact that python first does the `is` operator and then the `==`.

Related numpy issue: https://github.com/numpy/numpy/issues/9358
